src/core/mac/mac.cpp

```diff
@@ -2012,5 +2136,225 @@ uint8_t Mac::GetTimeIeOffset(const Frame &aFrame)
 }
 #endif
 
+#if OPENTHREAD_CONFIG_CSL_RECEIVER_ENABLE
+void Mac::SetCslChannel(uint8_t aChannel)
+{
+    VerifyOrExit(GetCslChannel() != aChannel);
+
+    mSubMac.SetCslChannel(aChannel);
+
+    if (mSubMac.IsCslStarted())
+    {
+        Get<Mle::Mle>().ScheduleChildUpdateRequest();
+    }
+exit:
+    return;
+}
+
+void Mac::SetCslPeriod(uint16_t aPeriod)
+{
+    VerifyOrExit(GetCslPeriod() != aPeriod);
+
+    mSubMac.SetCslPeriod(aPeriod);
+
+    if (aPeriod == 0)
+    {
+        StopCsl();
+    }
+    else if (mSubMac.IsCslStarted())
+    {
+        Get<Mle::Mle>().ScheduleChildUpdateRequest();
+    }
+    else
+    {
+        StartCsl();
+    }
+
+exit:
+    return;
+}
+
+void Mac::SetCslTimeout(uint32_t aTimeout)
+{
+    VerifyOrExit(GetCslTimeout() != aTimeout);
+
+    mSubMac.SetCslTimeout(aTimeout);
+
+    if (mSubMac.IsCslStarted())
+    {
+        Get<Mle::Mle>().ScheduleChildUpdateRequest();
+    }
+exit:
+    return;
+}
+
+bool Mac::ShouldIncludeCslIe(void) const
+{
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    return mSubMac.IsCslStarted() && !IsCslIeSuppressed() &&
+           !Get<Mle::Mle>().IsRxOnWhenIdle(); // Only for certification
+#else
+    return mSubMac.IsCslStarted() && !Get<Mle::Mle>().IsRxOnWhenIdle();
+#endif // OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+}
+
+otError Mac::StartCsl(void)
+{
+    otError   error  = OT_ERROR_NONE;
+    Neighbor *parent = &Get<Mle::Mle>().GetParent();
+
+    VerifyOrExit(!mRxOnWhenIdle && Get<Mle::MleRouter>().GetRole() == OT_DEVICE_ROLE_CHILD,
+                 error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(parent->IsSupportEnhancedFramePending(), error = OT_ERROR_NOT_CAPABLE);
+
+    otPlatRadioEnableCsl(&GetInstance(), mSubMac.GetCslPeriod(), &parent->GetExtAddress());
+
+    SuccessOrExit(error = mSubMac.StartCsl());
+
+    Get<Mle::Mle>().ScheduleChildUpdateRequest();
+
+exit:
+    return error;
+}
+
+void Mac::StopCsl(void)
+{
+    Neighbor *parent = &Get<Mle::Mle>().GetParent();
+    VerifyOrExit(parent != NULL);
+
+    VerifyOrExit(mSubMac.IsCslStarted()); // This is required to avoid StopCsl is called during initialization and call
+                                          // otPlatRadioEnableCsl
+
+    mSubMac.StopCsl();
+
+    otPlatRadioEnableCsl(&GetInstance(), 0, &parent->GetExtAddress());
+
+exit:
+    return;
+}
+#endif // OPENTHREAD_CONFIG_CSL_RECEIVER_ENABLE
+
+#if OPENTHREAD_CONFIG_CSL_TRANSMITTER_ENABLE
+void Mac::ProcessCsl(const RxFrame &aFrame, const Address &aSrcAddr)
+{
+    const uint8_t *cur   = aFrame.GetHeaderIe(Frame::kHeaderIeCsl);
+    Child *        child = Get<ChildTable>().FindChild(aSrcAddr, Child::kInStateAnyExceptInvalid);
+    const CslIe *  csl   = reinterpret_cast<const CslIe *>(cur + sizeof(HeaderIe));
```

Comment:
Is it okay to perform arithmetic operation here with a possible NULL pointer? Are we okay with undefined behavior?

Reply:
I think it's ok to do arithmetic operation of `NULL` pointer. It's just like adding something to `0`. So long as we don't access the memory it points to, it won't cause some problems I think. And we do the verification in the next line.

Reply:
OK.